src/main/java/pl/edu/pw/mini/ingreedio/api/service/ProductsCriteriaService.java

```diff
@@ -0,0 +1,123 @@
+package pl.edu.pw.mini.ingreedio.api.service;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+import pl.edu.pw.mini.ingreedio.api.criteria.ProductCriteria;
+import pl.edu.pw.mini.ingreedio.api.criteria.ProductsSortingCriteria;
+import pl.edu.pw.mini.ingreedio.api.dto.IngredientDto;
+
+@Service
+@AllArgsConstructor
+public class ProductsCriteriaService {
+    private final IngredientService ingredientService;
+
+    public ProductCriteria getProductsCriteria(Optional<List<Long>> ingredientsToExclude,
+                                               Optional<List<Long>> ingredientsToInclude,
+                                               Optional<Integer> minRating,
+                                               Optional<String> phrase,
+                                               Optional<List<String>> sortBy,
+                                               Optional<Boolean> liked) {
+
+        // TODO: provider, brand, category (add arguments too)
+        var builder = ProductCriteria.builder();
+        builder.hasMatchScoreSortCriteria(false);
+
+        ingredientsToExclude.ifPresent(
+            ingredients -> {
+                Set<Long> ingredientsSet = new HashSet<>(ingredients);
+                builder.ingredientsNamesToExclude(ingredientService
+                    .getIngredientsByIds(ingredientsSet)
+                    .stream()
+                    .map(IngredientDto::name)
+                    .collect(Collectors.toSet()));
+            });
+
+        ingredientsToInclude.ifPresent(
+            ingredients -> {
+                Set<Long> ingredientsSet = new HashSet<>(ingredients);
+                builder.ingredientsNamesToInclude(ingredientService
+                    .getIngredientsByIds(ingredientsSet)
+                    .stream()
+                    .map(IngredientDto::name)
+                    .collect(Collectors.toSet()));
+            });
+
+        minRating.ifPresent(builder::minRating);
+
+        phrase.ifPresent(phraseString ->
+            builder.phraseKeywords(Arrays.stream(
+                phraseString.replaceAll("%20+", " ")
+                    .trim()
+                    .split(" "))
+                .collect(Collectors.toSet())));
+
+        liked.ifPresent(builder::liked);
+
+        sortBy.ifPresent(sortingSignatures -> {
+            List<ProductsSortingCriteria> sortingCriteriaList =
+                sortingSignatures.stream()
+                .map(this::getProductsSortingCriteria)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .toList();
```

Comment:
[Item 24: Do We Need to Chain the Optional API With the Stream API?](https://dzone.com/articles/using-optional-correctly-is-not-optional#:~:text=Item%2024%3A%20Do%20We%20Need%20to%20Chain%20the%20Optional%20API%20With%20the%20Stream%20API%3F)
```suggestion
            List<ProductsSortingCriteria> sortingCriteriaList =
                sortingSignatures.stream()
                .map(this::getProductsSortingCriteria)
                .flatMap(Optional::stream)
                .toList();
```